### PR TITLE
Adding Customer Usage Attribution to ESLZ/ALZ ARM and Portal template

### PR DESCRIPTION
**Pull request by jhajduk-microsoft**
## Overview/Summary

This pull request represents changes requested in backlog item 98657 - Implement a UUID/GUID in ESLZ template

## This PR fixes/adds/changes/removes

1. Add the Customer Usage Attribution feature to the ESLZ/ALZ template
2. Enhances the ESLZ/ALZ portal template with Customer Usage Attribution options

### Breaking Changes

1. No Breaking Changes

## Testing Evidence
(https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fjhajduk-microsoft%2FEnterprise-Scale%2Fmain%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fjhajduk-microsoft%2FEnterprise-Scale%2Fmain%2FeslzArm%2Feslz-portal.json)

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Enterprise-Scale/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/Enterprise-Scale/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Enterprise-Scale/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.
- [x] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located: `/docs/wiki/whats-new.md`)

### CODE REVIEW

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+To disable this tracking, we have included a parameter called `telemetryOptOut` to the ESLZ ARM Template in this repo with a simple boolean flag. The default value `false` which **does not** disable the telemetry. If you would like to disable this tracking, then simply set this value to `true` and this module will not be included in deployments and **therefore disables** the telemetry tracking. The portal experience ESLZ ARM Template has a radio button to toggle this option on/off.
+
+If you are happy with leaving telemetry tracking enabled, no changes are required. Please do not edit the module name or value of the variable `cuaID` in any module.
+
+For example, in the eslzARM.json file, you will see the following:
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:11]:
```suggestion
For example, in the `eslzARM.json` file, you will see the following:
```

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+[ESLZ ARM Template parameter example](./media/cua-parameter.png)
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:13]:
```suggestion
![ESLZ ARM Template parameter example](./media/cua-parameter.png)
```

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+[ESLZ ARM Template variable example](./media/cua-variable.png)
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:14]:
```suggestion
![ESLZ ARM Template variable example](./media/cua-variable.png)
```

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+[ESLZ ARM Template resource example](./media/cua-resource.png)
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:15]:
```suggestion
![ESLZ ARM Template resource example](./media/cua-resource.png)
```

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:16]:
```suggestion
```

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<!-- markdownlint-disable -->
+## Telemetry Tracking Using Customer Usage Attribution (PID)
+<!-- markdownlint-restore -->
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:3]:
```suggestion
```

not required in this repo

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<!-- markdownlint-disable -->
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:1]:
```suggestion
```
not required in this repo

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+## Telemetry Tracking Using Customer Usage Attribution (PID)
```
Comment (jtracey93) [inline docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:2]: Need to add this file into the following locations as links:

1. Wiki Sidebar 
2. Wiki Home
3. Repo root README.md

File: eslzArm/eslz-portal.json [hunk 9]
```diff
@@ -108,7 +108,37 @@
...
 							"visible": "[equals(steps('lzSettings').subOrgsOption, 'Single')]"
-            			}
+            			},
+                        {
+                            "name": "cuaSettingsInfo",
+                            "type": "Microsoft.Common.InfoBox",
+                            "visible": true,
+                            "options": {
+                                "text": "See Enterprise Scale Repository Docs > Wiki > Deploying-Enterprise-Scale-CustomerUsage for complete info. Microsoft can identify the deployments of the Azure Resource Manager templates with the deployed Azure resources. The usage of this data is governed by Microsoft Trust and Privacy policies. The customer usage identifier is used for telemetry purposes.",
+                                "uri": "https://github.com/Azure/Enterprise-Scale/blob/main/docs/EnterpriseScale-CustomerUsage.md",
```
Comment (jtracey93) [inline eslzArm/eslz-portal.json:118]: Is this URL correct? Should point to Wiki, no?

File: eslzArm/eslzArm.json [hunk 10]
```diff
@@ -617,7 +617,21 @@
                 "Audit",
                 "No"
             ]
+        },
+        "telemetryOptOut": {
+            "type": "string",
+            "defaultValue": [
```
Comment (jtracey93) [inline eslzArm/eslzArm.json:623]: This should be a string, not an array?

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Nice work @jhajduk-microsoft, a few changes please 👍

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md (hunk 8 above)
Comment (jhajduk-microsoft) [reply docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:1]: Added

File: eslzArm/eslz-portal.json (hunk 9 above)
Comment (jhajduk-microsoft) [reply eslzArm/eslz-portal.json]: corrected URL

**Comment by jtracey93** [comment]
@jhajduk-microsoft is this good for another review? Let me know 👍👍

**Comment by jhajduk-microsoft** [comment]
@jtracey93 - good for review

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Looking great here @jhajduk-microsoft 👍

I have made a few changes in your branch to fix a few formatting and indenting issues. I have also added a section to the portal experience for the CUA piece to make it display a little nicer (breaking it up a little)
image

I have also left a few comments to be resolved elsewhere relating to docs etc. 

Look forward to getting this merged soon once you have updated based on remaining comments and re-tested 👍👍

Thanks

Jack

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+Microsoft can identify the deployments of the Azure Resource Manager templates with the deployed Azure resources. Microsoft can correlate these resources used to support the deployments. Microsoft collects this information to provide the best experiences with their products and to operate their business. The telemetry is collected through [customer usage attribution](https://docs.microsoft.com/azure/marketplace/azure-partner-customer-usage-attribution). The data is collected and governed by Microsoft's privacy policies, located at the [trust center](https://www.microsoft.com/trustcenter).
+
+To disable this tracking, we have included a parameter called `telemetryOptOut` to the ESLZ ARM Template in this repo with a simple boolean flag. The default value `false` which **does not** disable the telemetry. If you would like to disable this tracking, then simply set this value to `true` and this module will not be included in deployments and **therefore disables** the telemetry tracking. The portal experience ESLZ ARM Template has a radio button to toggle this option on/off.
+
```
[the remark this answers is not in the document]
Comment (jhajduk-microsoft) [reply docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:6]: Will do

[the remark this answers is not in the document]
Comment (jhajduk-microsoft) [reply docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md:6]: I rearranged the wiki a bit so it flows a bit better, will PR by EOD

**Comment by jhajduk-microsoft** [comment]
@jtracey93 - Ready for review
https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fjhajduk-microsoft%2FEnterprise-Scale%2Fmain%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fjhajduk-microsoft%2FEnterprise-Scale%2Fmain%2FeslzArm%2Feslz-portal.json

**Review by jtracey93 — APPROVED** [review]
Great work @jhajduk-microsoft. LGTM 👍

Going to request @krowlandson also reviews before we merge, as 2 pairs of eyes is better than one for a change like this 👍

I have tested deployment of this in the following scenarios:

1. Deployment with telemetry disabled 
image
2. Deployment over the top of same deployment with telemetry enabled
image
3. Deployment over the top of that same deployment with telemetry enabled
image

All worked fine and as expected 👍

**Review by krowlandson — CHANGES_REQUESTED** [review]
Looking good. Just a couple of minor feedback comments which would be good to consider.

**Comment by jtracey93** [comment]
@jhajduk-microsoft if you can resolve the comments from @krowlandson's review (https://github.com/Azure/Enterprise-Scale/pull/938#pullrequestreview-920382384) then I think we are good to get this merged 👍

File: docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
+## Telemetry Tracking Using Customer Usage Attribution (PID)
+
+Microsoft can identify the deployments of the Azure Resource Manager templates with the deployed Azure resources. Microsoft can correlate these resources used to support the deployments. Microsoft collects this information to provide the best experiences with their products and to operate their business. The telemetry is collected through [customer usage attribution](https://docs.microsoft.com/azure/marketplace/azure-partner-customer-usage-attribution). The data is collected and governed by Microsoft's privacy policies, located at the [trust center](https://www.microsoft.com/trustcenter).
+
+To disable this tracking, we have included a parameter called `telemetryOptOut` to the ESLZ ARM Template in this repo with a simple boolean flag. The default value `false` which **does not** disable the telemetry. If you would like to disable this tracking, then simply set this value to `true` and this module will not be included in deployments and **therefore disables** the telemetry tracking. The portal experience ALZ Accelerator/ESLZ ARM Deployment Template has a radio button to toggle this option on/off.
```
[the remark this answers is not in the document]
Comment (jhajduk-microsoft) [reply docs/wiki/Deploying-Enterprise-Scale-CustomerUsage.md]: Added wording to specify the portal experience as recommend and the ARM template experience as and additional option

**Comment by jhajduk-microsoft** [comment]
Changes are ready for review

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Thanks @jhajduk-microsoft, could we update the screenshots of the portal experience in the docs:
image

Thanks

Jack

**Comment by jhajduk-microsoft** [comment]
@jtracey93 - ready...this is the one!

**Review by jtracey93 — APPROVED** [review]
LGTM. Great work @jhajduk-microsoft 

My testing of 3 scenarios in 2 different regions ensured the PID was unique based on location and prefix used and was also the same when the inputs were the same as per design.

image

**Pull request merged**